### PR TITLE
fix: track global phase in operators

### PR DESCRIPTION
**Pull request by plerouxanyon**
This ensures that the global phase is correctly taken into account in pre-defined operators.

Since it's required to make things work, this also updates `compare_kets` to correctly handle the global phase if the first element is zero and adds `compare_operators`.

This is required for #358, but it does not yet ensure that the transpilation

### CODE REVIEW

File: src/core/qobj.jl [hunk 1]
```diff
@@ -1766,3 +1777,66 @@ function genlaguerre(x, alpha, n)
     end
     return result
 end
+
+function get_canonical_global_phase(H::AbstractMatrix{Complex{T}})::T where {T<:Real}
```
Comment (fSosaRey) [inline src/core/qobj.jl:1781]: one more place we should use this function is inside `as_universal_gate()`

File: src/core/quantum_gate.jl [hunk 2]
```diff
@@ -1506,12 +1506,11 @@ Return the `Operator` which applies a ``π/2`` rotation about the ``Z`` axis.
 The `Operator` is defined as:
```
[the remark this answers is not in the document]
Comment (fSosaRey) [reply src/core/quantum_gate.jl:1506]: also on line 1518

File: src/core/quantum_gate.jl [hunk 3]
```diff
@@ -1599,6 +1598,22 @@ P(\\phi) = \\begin{bmatrix}
 """
 phase_shift(phi, T::Type{<:Complex} = ComplexF64) = DiagonalOperator(T[1.0, exp(im * phi)])
 
+"""
+    rotation_z(lambda)
+
+Return the `DiagonalOperator` that applies a rotation of `z` .
```
Comment (fSosaRey) [inline src/core/quantum_gate.jl:1604]: symmetric phase offset should be mentioned

File: test/qobj.jl [hunk 4]
```diff
@@ -321,3 +321,11 @@ end
         hspace_size_per_body,
     )
 end
+
+@testset "get_canonical_global_phase" begin
```
Comment (fSosaRey) [inline test/qobj.jl:325]: missing tests for non-zero phases

File: src/core/qobj.jl [hunk 5]
```diff
@@ -64,6 +64,23 @@ end
 
 Base.length(x::Ket) = Base.length(x.data)
 
+function get_canonical_global_phase(Ψ::Complex)::Real
+    return atan(imag(Ψ), real(Ψ))
+end
+
+function get_canonical_global_phase(ψ::Ket{Complex{T}})::T where {T<:Real}
+    # Use the first non-zero element to determine the global phase of the matrix
```
[the remark this answers is not in the document]
Comment (plerouxanyon) [reply src/core/qobj.jl:72]: Fixed

File: src/core/qobj.jl (hunk 1 above)
Comment (plerouxanyon) [reply src/core/qobj.jl:1781]: Updated

File: src/core/qobj.jl [hunk 6]
```diff
@@ -1766,3 +1777,66 @@ function genlaguerre(x, alpha, n)
...
+    end
+
+    # If the matrix is zero, then define the global phase as 0
+    return T(0)
+end
+
+
+"""
+
+    compare_operators(H_0::AbstractOperator,H_1::AbstractOperator)::Bool
```
[the remark this answers is not in the document]
Comment (plerouxanyon) [reply src/core/qobj.jl:1797]: Good catch, updated

File: src/core/quantum_gate.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (plerouxanyon) [reply src/core/quantum_gate.jl:1506]: I understand what you mean by "phase offset is symmetric", but I can't find it by googling. I've added that the unitary is symmetric, but "symmetric phase offset" sounds strange to me. Do you have a reference of it somewhere?

File: src/core/quantum_gate.jl (hunk 3 above)
Comment (plerouxanyon) [reply src/core/quantum_gate.jl:1604]: Same as above

File: test/qobj.jl (hunk 4 above)
Comment (plerouxanyon) [reply test/qobj.jl:325]: Added more tests

File: src/core/quantum_gate.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (fSosaRey) [reply src/core/quantum_gate.jl:1507]: No, I came up with that, which isn't really standard. Also, saying "the unitary is symmetric" is misleading as "symmetric" for matrices means `U^T = U`. We could put it as "the phase offset is applied symmetrically to the |0> and |1> states". Either that or remove it altogether. Qiskit doesnt make any mention of it, just gives the matrices.

[the remark this answers is not in the document]
Comment (plerouxanyon) [reply src/core/quantum_gate.jl:1506]: While the rotation gates are symmetric, that probably isn't useful information. I removed the mention of symmetry since we aren't really going into the properties of other operators.

**Review by fSosaRey — APPROVED** [review]
